### ISSUE TITLE
Phase 2a: YOLO Detection Foundation — replace template matching with YOLOv11

### ISSUE THREAD

**Issue by Armin2708** [Armin2708/ClashOfClans-Bot#1]
## Goal

Replace the current OpenCV template-matching vision system in `bot/vision.py` with a YOLOv11-based object detector that understands every element of the game: buildings, defenses, hero pads, troops, UI buttons, and screen state indicators.

## Background

The current system uses `cv2.matchTemplate` against a library of PNG templates. It is brittle — it breaks on resolution changes, game updates, or UI variations. YOLOv11 trained on real CoC screenshots will generalise far better and enable Phase 2b (intelligent attack decisions).

## Research findings

**Public baselines to start from:**
- [keremberke/clash-of-clans-object-detection](https://huggingface.co/datasets/keremberke/clash-of-clans-object-detection) — 125 images, 16 building/defense classes (cannon, mortar, inferno, xbow, eagle, scattershot, airsweeper, bombtower, clancastle, kingpad, queenpad, rcpad, wardenpad, wizztower, ad, th13), COCO format
- [WalkStation on Roboflow](https://universe.roboflow.com/walkstation/clash-of-clan-object-detection) — 798 images + pre-trained YOLOv8 model
- [Find This Base on Roboflow](https://universe.roboflow.com/find-this-base/clash-of-clans-vop4y/dataset/4) — augmented 640×640 variant, CC BY 4.0
- [pyAutoYoloCOC](https://github.com/ZhangRenL/pyAutoYoloCOC) — reference Python bot using YOLOv11 for CoC automation

**What public datasets are missing (must be captured ourselves):**
- UI buttons: attack, find_match, start_battle, next_base, return_home, end_battle
- Screen state indicators: village HUD, results/stars screen, searching animation, army screen
- Troops deployed in battle
- Loot values (gold/elixir text — currently handled by digit OCR templates)

### New files
- `bot/detector.py` — `Detector` class wrapping `ultralytics.YOLO`. Single responsibility: run inference on a frame and return `list[Detection]` (class, confidence, bbox).
- `training/download_datasets.py` — downloads public CoC datasets from Roboflow + HuggingFace, normalises to YOLOv11 format
- `training/capture_frames.py` — hooks into the running bot to save labelled screenshots automatically during attacks
- `training/train.py` — fine-tune YOLOv11n/s on merged dataset
- `training/dataset.yaml` — unified class list and dataset paths
- `models/` — `.pt` model weights (gitignored except the committed baseline)

### Modified files
- `bot/vision.py` — replace all `cv2.matchTemplate` calls with `Detector` queries. Keep the same public API (`find_button`, `detect_screen_state`, `read_enemy_loot`, `get_troop_slots`) so no callers change.
- `bot/settings.py` — add `yolo_model_path`, `yolo_confidence_threshold` settings
- `requirements.txt` — add `ultralytics>=8.3`

### Inference pipeline
```
screenshot (BGR np.ndarray)
  → Detector.predict(frame)           # YOLOv11 inference at 640px
  → list[Detection(cls, conf, bbox)]  # filtered by confidence threshold
  → find_button / detect_screen_state / get_troop_slots  # unchanged API
```

### Class list (target — superset of public datasets)
Buildings/defenses: `cannon, mortar, inferno, xbow, eagle, scattershot, airsweeper, bombtower, clancastle, wizztower, townhall, goldmine, elixirstorage, darkelixirdriller, builderhutt`
Heroes/pads: `kingpad, queenpad, wardenpad, rcpad`
UI buttons: `btn_attack, btn_find_match, btn_start_battle, btn_next_base, btn_return_home, btn_end_battle, btn_confirm`
Screen state indicators: `hud_village, hud_scouting, hud_results, hud_searching, hud_army`
Loot display: `loot_gold, loot_elixir`

## Data capture pipeline

`training/capture_frames.py` runs alongside the bot and saves a screenshot + auto-generated partial label (from current screen state + any confirmed button tap) every N seconds. Labels are saved in YOLOv11 TXT format for easy review in LabelImg or Roboflow.

## Implementation stages

1. Download public datasets, merge, train baseline YOLOv11n model
2. Evaluate on our own 2560×1440 BlueStacks screenshots — identify gaps
3. Capture 200–500 frames from live bot runs covering all game states
4. Label the gap classes (UI buttons, screen states)
5. Fine-tune merged dataset → new model weights
6. Implement `bot/detector.py` with `Detector` class
7. Rewrite `bot/vision.py` internals to use `Detector`, keep public API unchanged
8. Integration test: run the bot end-to-end and verify all states detected correctly
9. Commit final model weights + training scripts

## Success criteria

- Bot completes a full attack cycle (`do_attack()`) using only YOLO detection with no template files
- `detect_screen_state` correctly classifies all 7 game states in integration test
- `find_button` finds all 6 buttons with ≥ 0.80 confidence on 20 test screenshots
- `get_troop_slots` identifies troop slots during battle
- Inference latency ≤ 50ms per frame on CPU (MacBook M-series)

## Depends on
Nothing — this is the Phase 2 starting point.

## Blocks
Phase 2b: Intelligent Attack Engine (#next-issue)

## Labels
`enhancement`, `phase-2`, `ml`

**Comment by Armin2708** [comment]
Superseded by #3, which contains the complete executable implementation plan.
